com.unity.render-pipelines.universal/Shaders/LitDisplacement.hlsl

```diff
@@ -0,0 +1,44 @@
+#ifndef UNIVERSAL_LIT_DISPLACEMENT_INCLUDED
+#define UNIVERSAL_LIT_DISPLACEMENT_INCLUDED
+
+// Return view direction in tangent space, make sure tangentWS.w is already multiplied by GetOddNegativeScale()
+half3 GetViewDirectionTangentSpace(half4 tangentWS, half3 normalWS, half3 viewDirWS)
+{
+    // must use interpolated tangent, bitangent and normal before they are normalized in the pixel shader.
+    half3 unnormalizedNormalWS = normalWS;
+    const half renormFactor = 1.0 / length(unnormalizedNormalWS);
+
+    // use bitangent on the fly like in hdrp
+    // IMPORTANT! If we ever support Flip on double sided materials ensure bitangent and tangent are NOT flipped.
+    half crossSign = (tangentWS.w > 0.0 ? 1.0 : -1.0); // we do not need to multiple GetOddNegativeScale() here, as it is done in vertex shader
+    half3 bitang = crossSign * cross(normalWS.xyz, tangentWS.xyz);
+
+    half3 WorldSpaceNormal = renormFactor * normalWS.xyz;		// we want a unit length Normal Vector node in shader graph
+
+    // to preserve mikktspace compliance we use same scale renormFactor as was used on the normal.
+    // This is explained in section 2.2 in "surface gradient based bump mapping framework"
+    half3 WorldSpaceTangent = renormFactor * tangentWS.xyz;
+    half3 WorldSpaceBiTangent = renormFactor * bitang;
+
+    half3x3 tangentSpaceTransform = half3x3(WorldSpaceTangent, WorldSpaceBiTangent, WorldSpaceNormal);
+    half3 viewDirTS = mul(tangentSpaceTransform, viewDirWS);
+
+    return viewDirTS;
+}
+
+half2 ParallaxOffset1Step(half height, half amplitude, half3 viewDirTS)
```

Comment:
maybe this function should live in shader library?

Reply:
from API standpoint It seems a little bit misleading that the function expects unormalized viewDirTS, it might lead to double normalization as I'd think at least to pass any dir argument normalized.

Reply:
Fixed, now its in com.unity.render-pipelines.core\ShaderLibrary\ParallaxMapping.hlsl